webapp/src/ts/modules/contacts/contacts-edit.component.ts

```diff
@@ -0,0 +1,318 @@
+      if (_isEqual(this.routeSnapshot.params, params)) {
+        // the 1st time we load the form, we must wait for the view to be initialized
+        // if we don't skip, it will result in the form being loaded twice
+        return;
+      }
+      this.routeSnapshot = this.route.snapshot;
+
+      this.resetState();
+      this.resetFormError();
+      this.initForm();
+    });
+    this.subscription.add(routeSubscription);
+
+    const queryParamsSubscription = this.route.queryParams.subscribe(() => {
+      this.routeSnapshot = this.route.snapshot;
+      this.setCancelCallback();
+    });
+    this.subscription.add(queryParamsSubscription);
+  }
+
+  private resetState() {
+    this.globalActions.setLoadingContent(true);
+    this.globalActions.setShowContent(true);
+
+    if (!this.routeSnapshot.params?.id) {
+      this.globalActions.unsetSelected();
+      this.globalActions.settingSelected();
+    }
+  }
+
+  private setCancelCallback() {
+    this.globalActions.setCancelCallback(() => {
+      if (this.routeSnapshot.queryParams?.from === 'list') {
+        this.router.navigate(['/contacts']);
+      } else {
+        const parentContactId = this.routeSnapshot.params.id || this.routeSnapshot.params.parent_id;
+        this.router.navigate(['/contacts', parentContactId]);
+      }
+    });
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+    this.translationsLoadedSubscription?.unsubscribe();
+    this.globalActions.setTitle();
+    if (this.enketoContact?.formInstance) {
+      this.enketoService.unload(this.enketoContact.formInstance);
+    }
+  }
+
+  ngAfterViewInit() {
+    this.initForm();
+  }
+
+  private initForm() {
+    this.contentError = false;
+    this.errorTranslationKey = false;
+
+    return this
+      .getContact()
+      .then((contact) => this.getForm(contact))
+      .then((formId) => this.renderForm(formId))
+      .then((formInstance) => this.setEnketoContact(formInstance))
+      .then(() => {
+        this.globalActions.setLoadingContent(false);
+      })
+      .catch((err) => {
+        this.errorTranslationKey = err.translationKey || 'error.loading.form';
+        this.globalActions.setLoadingContent(false);
+        this.contentError = true;
+        console.error('Error loading contact form.', err);
+      });
+  }
+
+  private getFormInstanceData() {
+    const type = this.contact?.contact_type || this.contact?.type;
+    if (!type) {
+      return null;
+    }
+
+    return { [type]: this.contact };
+  }
+
+  private getContact() {
+    const id = this.routeSnapshot.params.id;
+    if (!id) {
+      return Promise.resolve();
+    }
+    return this.lineageModelGeneratorService
+      .contact(id, { merge: true })
+      .then((result) => result.doc);
+  }
+
+  private getForm(contact) {
+    let formId;
+    let titleKey;
+    const typeId = contact?.contact_type || contact?.type || this.routeSnapshot.params?.type;
+    return this.contactTypesService.get(typeId).then(type => {
```

Comment:
```suggestion
    return this.contactTypesService
      .get(typeId)
      .then(type => {
```
Also, we can `await` so we can flat the statements below by removing the `then`

Reply:
I don't like mixing async/await with promises in the same file, it feels messy. So I'm not switching to async/await because that would warrant a complete rewrite, which is risky.